### ISSUE TITLE
Use `.htaccess` files to secure configuration directories

### ISSUE THREAD

**Issue by ghost** [backdrop/backdrop-issues#706]
Currently, the `config` directory is 'secured' by appending a hash string to the dir path. This has caused some issues when changing database credentials, etc., and could be improved.

I suggest instead adding a .htaccess file to the directory upon creation (or when there isn't one, see below) that denies access in the same way that the .htaccess file in the `private` files directory works.

Also, when you delete the .htaccess file from the files directory and re-submit the File System form (admin/config/media/file-system), it re-creates the .htaccess file automatically. The same should be true of the `config` directory, and maybe even add a field to the File System page where you can specify the `config` directory (rather than in 'settings.php').

Advocate: @BWPanda

---

~Original PR: https://github.com/backdrop/backdrop/pull/3238~
New PR to address feedback form code review https://github.com/backdrop/backdrop/pull/3279

**Comment by BWPanda** [comment]
Forgot to mention, this is based on the discussion here: https://github.com/backdrop/backdrop-issues/issues/683

**Comment by quicksketch** [comment]
This sounds like a good idea to me. Additional security won't hurt.

> and maybe even add a field to the File System page where you can specify the config directory (rather than in 'settings.php').

We can't do this part however, since the config directory would have nowhere to store the config directory setting. :smile: That is, the config directory has to be in settings.php for the same reason why you wouldn't be able to store the database connection string in the database.

**Comment by BWPanda** [comment]
Made a PR that adds htaccess files to the config directories. It also checks they're working in `system_requirements()` and adds tests to make sure they're re-created if deleted.

**Comment by BWPanda** [comment]
Without this feature, you can access any config file like so: `http://[DOMAIN]/files/config_[HASH]/active/admin_bar.settings.json` That shows you the json for the Admin Bar config. Obviously the config dir needs to be under the Backdrop root dir somewhere (normally in `files`), and you need to know the hash. But since it's possible to set the config directories without a hash, then it's possible you don't need that.

With this feature, the same URL produces this:

> Forbidden
>
> You don't have permission to access this resource.

**Comment by indigoxela** [comment]
I appreciate this switch from "security by obscurity" to real security.

I tested a fresh install using your branch. Checked that the .htaccess files were created in both config dirs (active/staging). The files prevent access to these directories even if one knows the exact url. Works like a charm.

Then I manually deleted the file in the active dir and visited admin/reports/status. And the file has been recreated.

What I didn't actually test is a config sync.

**Comment by BWPanda** [comment]
Thanks @indigoxela. I will advocate for this feature.

**Comment by klonos** [comment]
> I appreciate this switch from "security by obscurity" to real security.

Well, I'd say "better security" ...this assumes apache, as nginx unfortunately(?) doesn't "talk" .htaccess: https://www.nginx.com/resources/wiki/start/topics/examples/likeapache-htaccess

**Comment by domaingood** [comment]
I think File permission will be good security Protect config directories With 500 and File 444 And More better If we can keep outside for www / Public_html with File permission

**Comment by indigoxela** [comment]
@domaingood It is already possible to keep config directories *outside* webroot without any problem. Ask in the forum or chat, if you're interested and need instructions (it's a bit off-topic here).

Regarding the file permissions you suggest: Backdrop needs to be able to *write* config, 444 for files would prevent that - don't do it - you'll get a ConfigStorageException.

**Comment by domaingood** [comment]
@indigoxela Thank you for reply.
But I think it need to Automated like Drupal Change their Directory 500 and setting files 444 When installing Drupal 7.
Also Make it config directory 500 and Files are 444 Automated So We are safe hand. If possible to move to outside of Public_html /www It will make more security our Backdrop CMS

**Comment by indigoxela** [comment]
> Directory 500 and setting files 444 When installing Drupal 7...

Now I got it - that's a complete misunderstanding. Here in this Issue "config directory" means where the various json files for the [Configuration Management System](https://api.backdropcms.org/documentation/working-configuration) get stored, _not_ to get confused with "sites/default", where Drupal 7 stores its settings.php file. Drupal 7 has no CMI.

**Comment by domaingood** [comment]
I m not confused With  "sites/default", where Drupal 7 stores its settings.php  I Want backdrop Doing same things with their JSON files  and Directories. So It will be secure.

**Comment by quicksketch** [comment]
Hi @domaingood, I remember you brought up file permissions for the config directories during a weekly meeting. It's something that we can evaluate separately from this discussion. Can you make a new issue? Adding .htaccess to the config directory is something we want to do in any case.

**Comment by domaingood** [comment]
I personally disagree .htaccess Because Nginx Don't Allow .htaccess. If added .htaccess its a fully Solution.

**Comment by quicksketch** [comment]
It's true this protection is only applicable if you're using Apache. The code checks the web server and only adds .htaccess files if you're running Apache. So for users of Nginx, Backdrop doesn't put .htaccess files anywhere because they're not going to do anything.

I posted a review to https://github.com/backdrop/backdrop/pull/3238#pullrequestreview-477011124, this looks really good and simple.

**Comment by BWPanda** [comment]
Don't think I can get to this, so someone else feel free to take over.

**Comment by indigoxela** [comment]
The [sandbox](https://github.com/backdrop/backdrop/pull/3279) seems broken, can someone with proper permissions restart it, please?

It's currently hard to tell if the tests failed because of a timeout or because they don't work properly yet.

**Comment by jenlampton** [comment]
The error message on the status report might need an update for these config directories. The current message only applies to the files directory: `See http://drupal.org/SA-CORE-2013-003 for information about the recommended .htaccess file which should be added to the ./files/config_b689320d98083ab1f4ad1386fc2cc3e6/active directory to help protect against arbitrary code execution.`

**Comment by indigoxela** [comment]
@jenlampton I left two helpful comments regarding the failing tests.

Unfortunately this does not solve the problem with the file not created in initializeStorage() - only in the test. I checked with aborting a test run - there really was no htaccess file in "files/simpletest/294561/config_active/".

**Comment by jenlampton** [comment]
Thanks @indigoxela for the review, I've made those changes and now we're down to 2 failing tests (an improvement!). I'm still not sure why the `.htaccess` file is created properly in normal usage, but not during the test. I don't think it's a permissions issue on the test config location, since that would prevent config from working entirely. Anyone have any ideas?

**Comment by quicksketch** [comment]
I pushed an update to https://github.com/backdrop/backdrop/pull/3279 that should fix the failing test.

Indeed everything was working perfectly, but the test was failing because SimpleTest runs a special process to emulate installation in between tests. We needed to update `BackdropWebTestCase::prepareEnvironment()` to have it initialize the config storage in a way that matches the normal install process. After changing this, the one failing test now passes because the needed .htaccess file is present immediately after SimpleTest runs the installation.
